irc/histserv.go

```diff
@@ -0,0 +1,242 @@
+			enabled:   histservEnabled,
+			minParams: 1,
+			maxParams: 2,
+		},
+	}
+)
+
+// histNotice sends the client a notice from HistServ
+func histNotice(rb *ResponseBuffer, text string) {
+	rb.Add(nil, histServMask, "NOTICE", rb.target.Nick(), text)
+}
+
+func histservForgetHandler(server *Server, client *Client, command string, params []string, rb *ResponseBuffer) {
+	accountName := server.accounts.AccountToAccountName(params[0])
+	if accountName == "" {
+		histNotice(rb, client.t("Could not look up account name, proceeding anyway"))
+		accountName = params[0]
+	}
+
+	server.ForgetHistory(accountName)
+
+	histNotice(rb, fmt.Sprintf(client.t("Enqueued account %s for message deletion"), accountName))
+}
+
+func histservDeleteHandler(server *Server, client *Client, command string, params []string, rb *ResponseBuffer) {
+	var target, msgid string
+	if len(params) == 1 {
+		msgid = params[0]
+	} else {
+		target, msgid = params[0], params[1]
+	}
+
+	accountName := "*"
+	hasPrivs := client.HasRoleCapabs("history")
+	if !hasPrivs {
+		accountName = client.AccountName()
+		if !(server.Config().History.Retention.AllowIndividualDelete && accountName != "*") {
+			hsNotice(rb, client.t("Insufficient privileges"))
+			return
+		}
+	}
+
+	err := server.DeleteMessage(target, msgid, accountName)
+	if err == nil {
+		hsNotice(rb, client.t("Successfully deleted message"))
+	} else {
+		if hasPrivs {
+			hsNotice(rb, fmt.Sprintf(client.t("Error deleting message: %v"), err))
+		} else {
+			hsNotice(rb, client.t("Could not delete message"))
+		}
+	}
+}
+
+func histservExportHandler(server *Server, client *Client, command string, params []string, rb *ResponseBuffer) {
+	cfAccount, err := CasefoldName(params[0])
+	if err != nil {
+		histNotice(rb, client.t("Invalid account name"))
+		return
+	}
+
+	config := server.Config()
+	filename := fmt.Sprintf("%s@%s.json", cfAccount, time.Now().UTC().Format(IRCv3TimestampFormat))
+	pathname := config.getOutputPath(filename)
+	outfile, err := os.Create(pathname)
+	if err != nil {
+		hsNotice(rb, fmt.Sprintf(client.t("Error opening export file: %v"), err))
+	} else {
+		hsNotice(rb, fmt.Sprintf(client.t("Started exporting account data to file %s"), pathname))
+	}
+
+	go histservExportAndNotify(server, cfAccount, outfile, client.Nick())
+}
+
+func histservExportAndNotify(server *Server, cfAccount string, outfile *os.File, alertNick string) {
+	defer func() {
+		if r := recover(); r != nil {
+			server.logger.Error("history",
+				fmt.Sprintf("Panic in history export routine: %v\n%s", r, debug.Stack()))
+		}
+	}()
+
+	defer outfile.Close()
+	writer := bufio.NewWriter(outfile)
+	defer writer.Flush()
+
+	server.historyDB.Export(cfAccount, writer)
+
+	client := server.clients.Get(alertNick)
+	if client != nil && client.HasRoleCapabs("history") {
+		client.Send(nil, histServMask, "NOTICE", client.Nick(), fmt.Sprintf(client.t("Data export for %[1]s completed and written to %[2]s"), cfAccount, outfile.Name()))
```

Comment:
Is there any concern here with giving the full file path? shouldn't be a drama but may be cleaner to e.g. split by `\` and `/` and return only the last portion or something